File: chapter15/RowPrefetching.cs
/**************************************************************************
******************* Troubleshooting Oracle Performance ********************
************************* http://top.antognini.ch *************************
***************************************************************************

File name...: RowPrefetching.cs
Author......: Christian Antognini
Date........: March 2009
Description.: This program provides an example of implementing the row
              prefetching with ODP.NET.
Notes.......: The table T created with row_prefetching.sql must exist.
Parameters. : <user> <password> <service name>

You can send feedbacks or questions about this script to top@antognini.ch.

Changes:
DD.MM.YYYY Description
---------------------------------------------------------------------------

**************************************************************************/

using System;
using Oracle.DataAccess.Client;

namespace TOP
{
    class RowPrefetching
    {
        static void Main(string[] args)
        {
            String user;
            String password;
            String dataSource;

            if (args.GetLength(0) != 3)
            {
                Console.WriteLine("usage: RowPrefetching.exe <user> <password> <service name>");
                return;
            }
            else
            {
                user = args[0];
                password = args[1];
                dataSource = args[2];
            }

            String connectString = "User Id=" + user + ";Password=" + password + ";Data Source=" + dataSource;
            OracleConnection connection = new OracleConnection(connectString);

            try
            {
                connection.Open();

                for (int i = 1; i <= 100; i++)
                {
                    DateTime startTime = DateTime.Now;
                    test(connection, i);
                    DateTime stopTime = DateTime.Now;
                    Console.WriteLine(i + " , " + (stopTime - startTime));
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.Message);
            }
            finally
            {
                if (connection.State == System.Data.ConnectionState.Open)
                {
                    connection.Close();
                    connection.Dispose();
                }
            }
        }

        private static void test(OracleConnection connection, int fetchSize)
        {
            String sql;
            OracleCommand command;
            OracleDataReader reader;
            Decimal id;
            String pad;

            sql = "SELECT id, pad FROM t";
            command = new OracleCommand(sql, connection);
            command.AddToStatementCache = false;
            reader = command.ExecuteReader();
            reader.FetchSize = command.RowSize * fetchSize;
            while (reader.Read())
            {
                id = reader.GetDecimal(0);
                pad = reader.GetString(1);
                // process data
                // Console.WriteLine(id + " , " + pad);
            }
            reader.Close();
        }
    }
}

File: chapter15/row_prefetching.php
<?php

/**************************************************************************
******************* Troubleshooting Oracle Performance ********************
************************* http://top.antognini.ch *************************
***************************************************************************

File name...: RowPrefetching.php
Author......: Christian Antognini
Date........: December 2013
Description.: This program provides an example of implementing the row
              prefetching with PHP.
Notes.......: The table T created with row_prefetching.sql must exist.
Parameters. : <user> <password> <service name>

You can send feedbacks or questions about this script to top@antognini.ch.

Changes:
DD.MM.YYYY Description
---------------------------------------------------------------------------

**************************************************************************/

  // definition of internal functions

  function get_parameter($name)
  {
    if (isset($_GET[$name]))
    {
      return $_GET[$name];
    }
    else
    {
      throw new Exception("parameter not defined");
    }
  }


  // get parameters

  try
  {
    $username = get_parameter("username");
    $password = get_parameter("password");
    $connect_string = get_parameter("connect_string");
    $fetch_size = get_parameter("fetch_size");
  }
  catch (Exception $e)
  {
    echo "<p>To run this script four parameters are needed:</p>";
    echo "<ul>";
    echo "<li>username</li>";
    echo "<li>password</li>";
    echo "<li>connect_string</li>";
    echo "<li>fetch_size</li>";
    echo "</ul>";
    echo "<p>E.g.: RowPrefetching.php?username=SCOTT&password=TIGER&connect_string=ORCL&fetch_size=42</p>";
    trigger_error("Invalid parameters", E_USER_ERROR);
  }

  // open connection

  $connection = oci_connect($username, $password, $connect_string);

  if (!$connection)
  {
    $error = oci_error();
    trigger_error($error['message'], E_USER_ERROR);
  }

  oci_set_client_identifier($connection, "cha");

  // test
 
  $sql = "SELECT id, pad FROM t";
  $statement = oci_parse($connection, $sql);
  oci_set_prefetch($statement, $fetch_size);
  oci_execute($statement, OCI_NO_AUTO_COMMIT);
  while ($row = oci_fetch_assoc($statement))
  {
    $id = $row['ID'];
    $pad = $row['PAD'];
    // process data
  }
  oci_free_statement($statement);

  // close connection

  if (!oci_close($connection))
  {
    $error = oci_error();
    trigger_error($error['message'], E_USER_ERROR);
  }

  echo "<p>execution successful</p>";

?>

